### ISSUE TITLE
[META] Create a "Backdrop Ready" module for D7 and generally make the switch from Drupal a more pleasurable process.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#2322]
So I was watching the training videos from BADCamp and it was clear to me that there can be a few issues when trying to upgrade a D7 test website to Backdrop. ??? I think that instead of expecting users to either already know what they need to do or to have them read documentation (that might put them off from upgrading and we don't want that), we should make the process as easy as possible for them. A D7 module that integrates with the site status page and would be perfect in those situations and there could be an equivalent core Backdrop module that would pick up after the upgrade was complete for any post-installation tasks. Once those tasks are either done or the user chooses to ignore as non-applicable to their situation, they can then safely disable that module. In simple words, we should have a D7 contrib and a respective Backdrop core module that "take users by the hand" and guide them through the process by doing some "helper" tasks for them:

- list all contrib modules/themes that the old site uses along with their status in Backdrop land. Something like [Upgrade Status](https://www.drupal.org/project/upgrade_status) but instead of checking Drupal current -> Drupal next, it would check D7 -> Backdrop. Then it would display the "audit" results in a nice status page:

 ![](https://www.drupal.org/files/images/upgrade_status.png)

 Something like that would help users decide if they are ready or not. Perhaps they can live without certain Drupal contrib modules if they seem to be 90% ready. Perhaps if most of their modules or at least the important ones are "in progress" (respective GitHub contrib repo created, but no official releases yet), then they'd feel more confident or urged to help with the porting.

- giving them "motivational" user-friendly stats in their site status page ("4 out of 5 pre-upgrade tasks complete. You are almost ready!").

- identify tasks that need to be done before the upgrade (disable Drupal contrib modules/themes, make a backup of their codebase and database etc.).

During the upgrade, we should be making educated guesses, but also provide options where applicable. For example, I saw during the video that there is a "Node layout" layout created with `node/%` as path. That's really intuitive (yay!!) and makes it possible to bring across any pre-existing visibility conditions (although #2321 happened during the training test upgrade). But we gave no option to the user to select a layout template for example or to place the rest of the blocks. OK, they can tweak that after the upgrade and I think that that is the best way to go about it (makes the upgrade a single step thing), but we did not tell them anything about it. Perhaps a "The blocks you had on your old site have been added to the *Node layout* layout. They should have been placed in appropriate regions, but perhaps you would like to double-check [link-to-manage-blocks-page-of-the-node-layout]."

Also at some point, @jenlampton mentioned for example that if you are using [Features](https://www.drupal.org/project/features) on the old site, then you need to first use [unfeaturize](https://www.drupal.org/project/unfeaturize) in order to get the content types out of features, and back into the database before moving on. We should have a hint about that and a link to that module.

Another thing was that if you are using a custom theme, it is advised to switch to core Bartik before upgrading. That would get you a working Backdrop version of your old site and then you could start either adding a completely new theme or porting the one you disabled. Again, another thing we should be giving a hint about (instead of the user going forward with the upgrade and discovering that they "broke their site" after the upgrade).

Anyways, these are all ideas and tasks that we should perhaps consider implementing in a way that makes the switch to Backdrop either be less research and work or at least help with these.

Thoughts? More ideas?

**Comment by jenlampton** [comment]
This is a fantastic idea! I remember something similar existing for Drupal already, was it coder_upgrade, maybe?

**Comment by serundeputy** [comment]
I also think this is a great idea!

> we should have a D7 contrib and a respective Backdrop core module that "take users by the hand" and guide them through the process

  * I think this would have to be a D7 module that lives on github.com/backdrop-contrib

> Also at some point, @jenlampton mentioned for example that if you are using Features on the old site, then you need to first use unfeaturize in order to get the content types out of features, and back into the database before moving on. We should have a hint about that and a link to that module if Features is detected on the old, to-be-upgraded site.

  * Is this really required?  Content types that are featurized are still in the database (they are just also in code).  When one performs an upgrade from Drupal 7 to Backdrop they have a Backdrop codebase downloaded and installed they then point the installed Backdrop at a backup of the D7 database (`backdrop-ready.sql`) it should pick up the content types no problem.  I will test this.

* Point being I think we should avoid cruft when trying to lead the users down the simples road to success.  So, I'm going to try the features thing and see if it is necessary.

As to all the other points that seem either necessary and/or very helpful to users I think make a backdrop-contrib repo `backdrop-ready` put the points you've made in the README.md and make issues for the other points (even if we no actual module code yet).  That should serve as a roadmap for devs to pluck off any or all of the work to be handled (you know becuase free time and what not).  

@klonos are you up for stubbing out this project in contrib?

I'll begin some features content type testing.

**Comment by serundeputy** [comment]
> Is this really required? Content types that are featurized are still in the database (they are just also in code). When one performs an upgrade from Drupal 7 to Backdrop they have a Backdrop codebase downloaded and installed they then point the installed Backdrop at a backup of the D7 database (backdrop-ready.sql) it should pick up the content types no problem. I will test this.

I've tested this and it is necessary.  Features must have the content types in the DB in a different way then standard ... not positive about that, but point is that `unfeaturizing` is not cruft it is a necessary step for `features` module users.

**Comment by klonos** [comment]
> I've tested this and it is necessary. Features must have the content types in the DB in a different way then standard ... not positive about that, but point is that `unfeaturizing` is not cruft it is a necessary step for `features` module users.

Precisely, and we're [talking 300k installations for D7](https://www.drupal.org/project/usage/features)! ...that's [as much as Views 7.x](https://www.drupal.org/project/usage/views)!!! That's why perhaps we should seriously consider including what unfeaturize.module does as part of our migration/upgrade scripts.

**Comment by klonos** [comment]
> As to all the other points that seem either necessary and/or very helpful to users I think make a backdrop-contrib repo backdrop-ready put the points you've made in the README.md and make issues for the other points (even if we no actual module code yet). That should serve as a roadmap for devs to pluck off any or all of the work to be handled (you know becuase free time and what not).

> @klonos are you up for stubbing out this project in contrib?

Almost bedtime now, but yes, I'll definitely get the repo and the README.md set up. I'll even file issues as to-dos in that repo so that anyone with the skills to get things implemented can have a go.

**Comment by klonos** [comment]
> I think this would have to be a D7 module that lives on github.com/backdrop-contrib

Would that make it so that once that project has releases they show up in the Project Browser (and b.org)? Do we want that if it is a D7 module?

**Comment by serundeputy** [comment]
We could just choose to not make releases of the `backdrop-ready` module, but reference it in certain upgrade path documentation.

**Comment by klonos** [comment]
I have made this issue be specific to the "Backdrop Ready" module(s), since most of the comments were around that. I have moved the more "generic" things to #3578 and made that issue a META.

**Comment by klonos** [comment]
Here's a screengrab of the upgrade_status module analysing a D8 site for D9 readiness: https://github.com/klonos/fileshare/blob/master/upgrade_status-D9-readiness-scan-2019-04-02-23.35.19.gif

**Comment by docwilmot** [comment]
Is anyone doing this?

**Comment by jenlampton** [comment]
I don't think anyone is working on it yet, but there is a Drupal 7 module called [upgrade_status](https://www.drupal.org/project/upgrade_status) that does the same thing for Drupal 8. That might be a good starting point?

**Comment by docwilmot** [comment]
https://github.com/backdrop-contrib/backdrop_upgrade_status/issues
